File: BM30_二叉搜索树与双向链表/javascript_solutions/StackSolution.ts
class TreeNode {
    val: number
    left: TreeNode | null
    right: TreeNode | null
    constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = (val === undefined ? 0 : val)
        this.left = (left === undefined ? null : left)
        this.right = (right === undefined ? null : right)
    }
}

/**
 * 代码中的类名、方法名、参数名已经指定，请勿修改，直接返回方法规定的值即可
 *
 * 
 * @param pRootOfTree TreeNode类 
 * @return TreeNode类
 */
export function Convert(pRootOfTree: TreeNode | null): TreeNode | null | undefined {
    if (!pRootOfTree) return null;

    let head: TreeNode | null | undefined = null;
    let pre: TreeNode | null | undefined = null;
    const stack: TreeNode[] = [];

    while (stack.length || pRootOfTree) {
        while (pRootOfTree) {
            stack.push(pRootOfTree);
            pRootOfTree = pRootOfTree.left;
        }

        const top = stack.pop();
        if (!top) throw new Error("出现神学事件，程序崩溃");

        if (!head || !pre) {
            head = top;
            pre = top;
        } else {
            pre.right = top;
            top.left = pre;
            pre = top;
        }

        pRootOfTree = top.right;
    }

    return head;
}

File: BM30_二叉搜索树与双向链表/python_solutions/stack_solution.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

#
# 
# @param pRootOfTree TreeNode类 
# @return TreeNode类
#
class Solution:
    def Convert(self, pRootOfTree ):
        if not pRootOfTree: return pRootOfTree

        head = None
        pre = None
        stack = []

        while pRootOfTree or stack:
            while pRootOfTree:
                stack.append(pRootOfTree)
                pRootOfTree = pRootOfTree.left
            
            top = stack.pop()

            if not head and not pre:
                head = top
                pre = top
            else:
                pre.right = top
                top.left = pre
                pre = top
            
            pRootOfTree = top.right
        
        return head

